### PR TITLE
Add Cuda SDK to Nvidia Docker image

### PR DESCRIPTION
**Pull request by ThomasRaoux**

### CODE REVIEW

File: build_tools/docker/nvidia/Dockerfile [hunk 1]
```diff
@@ -26,6 +26,15 @@ RUN wget -q "https://storage.googleapis.com/iree-shared-files/${NVIDIA_COMMON_DE
 RUN wget -q "https://storage.googleapis.com/iree-shared-files/${NVIDIA_GL_DEB?}"
 RUN wget -q "https://storage.googleapis.com/iree-shared-files/${NVIDIA_COMPUTE_DEB}"
 
+# Get the SDK from nvidia's repository
+RUN wget -q https://developer.download.nvidia.com/compute/cuda/repos/debian11/x86_64/cuda-cccl-11-6_11.6.55-1_amd64.deb
```
Comment (GMNGeoffrey) [inline build_tools/docker/nvidia/Dockerfile:30]: 😲 we really need all of these?

File: build_tools/docker/nvidia/Dockerfile [hunk 2]
```diff
@@ -45,8 +54,19 @@ COPY --from=fetch-nvidia \
   "/fetch-nvidia/${NVIDIA_COMMON_DEB}" \
   "/fetch-nvidia/${NVIDIA_GL_DEB}" \
   "/fetch-nvidia/${NVIDIA_COMPUTE_DEB}" \
+  "/fetch-nvidia/cuda-*" \
```
Comment (GMNGeoffrey) [inline build_tools/docker/nvidia/Dockerfile:57]: Huh this second build stage isn't really helping here anymore. I can futz with it though

File: build_tools/docker/prod_digests.txt [hunk 3]
```diff
@@ -6,6 +6,6 @@ gcr.io/iree-oss/frontends-nvidia@sha256:651e1c1c8dbe1cdfcca9648436e0c6fc9869d5df
 gcr.io/iree-oss/frontends-swiftshader@sha256:e70c6524cea980b9c6077a76e0ac90464c47eff4f27b6c74b8eaab3ff1fb35fc
 gcr.io/iree-oss/gradle-android@sha256:d36f3d172c3304e557cd3d9a13bae4590cb2a8d19d229b10f126423e7314a413
 gcr.io/iree-oss/riscv@sha256:24a401f9d47a16a0b460bf34a8485f087b7d88944ed661dd1201ce2d6c19df3e
-gcr.io/iree-oss/nvidia@sha256:b27c2feb5fc0e7125c5dd933c65c0f7029c157343c91269a17e892e76dd433a3
+gcr.io/iree-oss/nvidia@sha256:cad35c074dab074b6e36e2d5b4f36a2aba863d36013a93d50131db9bef66ec7f
```
Comment (GMNGeoffrey) [inline build_tools/docker/prod_digests.txt:9]: You likely need to also update the frontends-nvidia image as well

File: build_tools/kokoro/gcp_ubuntu/cmake/linux/x86-turing/build_kokoro.sh [hunk 4]
```diff
@@ -30,7 +30,7 @@ docker_setup
 
 docker run "${DOCKER_RUN_ARGS[@]?}" \
   --gpus all \
-  gcr.io/iree-oss/nvidia@sha256:b27c2feb5fc0e7125c5dd933c65c0f7029c157343c91269a17e892e76dd433a3 \
+  gcr.io/iree-oss/nvidia@sha256:cad35c074dab074b6e36e2d5b4f36a2aba863d36013a93d50131db9bef66ec7f \
```
Comment (GMNGeoffrey) [inline build_tools/kokoro/gcp_ubuntu/cmake/linux/x86-turing/build_kokoro.sh:33]: Let's enable cuda in these builds and tests that it works now

**Review by GMNGeoffrey — CHANGES_REQUESTED** [review]

File: build_tools/docker/nvidia/Dockerfile (hunk 1 above)
Comment (ThomasRaoux) [reply build_tools/docker/nvidia/Dockerfile:30]: nvcc-cuda depends on all those unfortunately

Comment (ThomasRaoux) [reply build_tools/docker/nvidia/Dockerfile]: changed it to a bigger package but that's the only thing I can make work with cmake

File: build_tools/docker/prod_digests.txt (hunk 3 above)
Comment (ThomasRaoux) [reply build_tools/docker/prod_digests.txt]: I'm only enabling nvidia-cmake build. We don't have tf integration tests with cuda so I don't think we need to right now and this avoid creating big dockers

File: build_tools/kokoro/gcp_ubuntu/cmake/linux/x86-turing/build_kokoro.sh (hunk 4 above)
Comment (ThomasRaoux) [reply build_tools/kokoro/gcp_ubuntu/cmake/linux/x86-turing/build_kokoro.sh]: Enabled the test for the x86-turing

**Comment by ThomasRaoux** [comment]
CUDA tests are passing with this version

File: build_tools/kokoro/gcp_ubuntu/cmake/linux/x86-turing/build.sh [hunk 5]
```diff
@@ -29,14 +29,66 @@ dpkg -l | grep nvidia
 nvidia-smi || true
 
 ./build_tools/kokoro/gcp_ubuntu/check_vulkan.sh
-
+IREE_SRC_DIR="$PWD"
 # TODO(gcmn): It would be nice to be able to build and test as much as possible,
 # so a build failure only prevents building/testing things that depend on it and
 # we can still run the other tests.
 # TODO: Add "-DIREE_TARGET_BACKEND_CUDA=ON -DIREE_HAL_DRIVER_CUDA=ON" once the
 # VMs have been updated with the correct CUDA SDK.
```
Comment (GMNGeoffrey) [inline build_tools/kokoro/gcp_ubuntu/cmake/linux/x86-turing/build.sh:37]: Please update the Todo here

File: build_tools/kokoro/gcp_ubuntu/cmake/linux/x86-turing/build.sh [hunk 6]
```diff
@@ -29,14 +29,66 @@ dpkg -l | grep nvidia
 nvidia-smi || true
 
 ./build_tools/kokoro/gcp_ubuntu/check_vulkan.sh
-
+IREE_SRC_DIR="$PWD"
```
Comment (GMNGeoffrey) [inline build_tools/kokoro/gcp_ubuntu/cmake/linux/x86-turing/build.sh:32]: This is redundant with the root dir below (and less robust)

File: build_tools/kokoro/gcp_ubuntu/cmake/linux/x86-turing/build.sh [hunk 7]
```diff
@@ -29,14 +29,66 @@ dpkg -l | grep nvidia
...
+
+  # Enable docs build on the CI. The additional builds are pretty fast and
+  # give us early warnings for some types of website publication errors.
+  "-DIREE_BUILD_DOCS=ON"
+
+  # Enable building the python bindings on CI. Most heavy targets are gated on
+  # IREE_ENABLE_TENSORFLOW, so what's left here should be fast.
+  "-DIREE_BUILD_PYTHON_BINDINGS=ON"
+
+  # Enable assertions. We wouldn't want to be testing *only* with assertions
```
Comment (GMNGeoffrey) [inline build_tools/kokoro/gcp_ubuntu/cmake/linux/x86-turing/build.sh:76]: This comment doesn't make sense here. I think having assertions here may be a good idea, but this is specifically referencing the script this was copied from

File: build_tools/kokoro/gcp_ubuntu/cmake/linux/x86-turing/build.sh [hunk 8]
```diff
@@ -29,14 +29,66 @@ dpkg -l | grep nvidia
...
+  "-DIREE_ENABLE_LLD=ON"
+  "-DCMAKE_BUILD_TYPE=RelWithDebInfo"
+
+  # Enable docs build on the CI. The additional builds are pretty fast and
+  # give us early warnings for some types of website publication errors.
+  "-DIREE_BUILD_DOCS=ON"
+
+  # Enable building the python bindings on CI. Most heavy targets are gated on
+  # IREE_ENABLE_TENSORFLOW, so what's left here should be fast.
+  "-DIREE_BUILD_PYTHON_BINDINGS=ON"
```
Comment (GMNGeoffrey) [inline build_tools/kokoro/gcp_ubuntu/cmake/linux/x86-turing/build.sh:74]: Do any cuda things require this setting?

File: build_tools/kokoro/gcp_ubuntu/cmake/linux/x86-turing/build.sh [hunk 9]
```diff
@@ -29,14 +29,66 @@ dpkg -l | grep nvidia
...
+
+CMAKE_ARGS=(
+  "-G" "Ninja"
+  # Let's make linking fast
+  "-DIREE_ENABLE_LLD=ON"
+  "-DCMAKE_BUILD_TYPE=RelWithDebInfo"
+
+  # Enable docs build on the CI. The additional builds are pretty fast and
+  # give us early warnings for some types of website publication errors.
+  "-DIREE_BUILD_DOCS=ON"
```
Comment (GMNGeoffrey) [inline build_tools/kokoro/gcp_ubuntu/cmake/linux/x86-turing/build.sh:70]: Definitely no reason to build these as part of a GPU build

File: build_tools/kokoro/gcp_ubuntu/cmake/linux/x86-turing/build.sh [hunk 10]
```diff
@@ -29,14 +29,66 @@ dpkg -l | grep nvidia
...
+
+ROOT_DIR=$(git rev-parse --show-toplevel)
+
+CMAKE_BIN=${CMAKE_BIN:-$(which cmake)}
+
+"$CMAKE_BIN" --version
+ninja --version
+
+cd ${ROOT_DIR?}
+if [ -d "build" ]
```
Comment (GMNGeoffrey) [inline build_tools/kokoro/gcp_ubuntu/cmake/linux/x86-turing/build.sh:53]: We just deleted the build dir above, so this check doesn't make sense

File: build_tools/docker/nvidia/Dockerfile [hunk 11]
```diff
@@ -50,3 +50,10 @@ COPY --from=fetch-nvidia \
 RUN apt-get install "/tmp/${NVIDIA_COMMON_DEB?}" \
   "/tmp/${NVIDIA_GL_DEB?}" \
   "/tmp/${NVIDIA_COMPUTE_DEB?}"
+
+# install cuda sdk
+RUN wget https://developer.download.nvidia.com/compute/cuda/11.6.0/local_installers/cuda-repo-debian11-11-6-local_11.6.0-510.39.01-1_amd64.deb
```
Comment (GMNGeoffrey) [inline build_tools/docker/nvidia/Dockerfile:55]: Oh good, I'm glad you figured out how to do this without a million packages

**Review by GMNGeoffrey — CHANGES_REQUESTED** [review]
It seems a bunch of files got marked executable for some reason? Could you please revert that

**Comment by ThomasRaoux** [comment]
> It seems a bunch of files got marked executable for some reason? Could you please revert that

Good catch, I fixed those and cleanup the script.

File: build_tools/kokoro/gcp_ubuntu/cmake/linux/x86-turing/build.sh (hunk 8 above)
Comment (ThomasRaoux) [reply build_tools/kokoro/gcp_ubuntu/cmake/linux/x86-turing/build.sh:60]: Turns out this is needed for the iree-external tests so I kept it.

File: build_tools/docker/prod_digests.txt (hunk 3 above)
Comment (GMNGeoffrey) [reply build_tools/docker/prod_digests.txt]: Ok, sure

File: build_tools/docker/nvidia/Dockerfile [hunk 12]
```diff
@@ -50,3 +50,10 @@ COPY --from=fetch-nvidia \
...
+RUN dpkg -i cuda-repo-debian11-11-6-local_11.6.0-510.39.01-1_amd64.deb
+RUN apt-key add /var/cuda-repo-debian11-11-6-local/7fa2af80.pub
+RUN apt-get update
+RUN apt-get -y install cuda-toolkit-11-6
```
Comment (GMNGeoffrey) [inline build_tools/docker/nvidia/Dockerfile:59]: I thought I commented on this before, but can't find it. Each `RUN` statement creates a new layer in the docker image, bloating the image. You also always want to have `apt-get update` as part of the same `RUN` statement as `apt-get install` (see https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#apt-get)

```suggestion
RUN wget https://developer.download.nvidia.com/compute/cuda/11.6.0/local_installers/cuda-repo-debian11-11-6-local_11.6.0-510.39.01-1_amd64.deb \
  && dpkg --install cuda-repo-debian11-11-6-local_11.6.0-510.39.01-1_amd64.deb \
  && apt-key add /var/cuda-repo-debian11-11-6-local/7fa2af80.pub \
  && apt-get update \
  && apt-get -y install cuda-toolkit-11-6
```

File: build_tools/docker/nvidia/Dockerfile [hunk 13]
```diff
@@ -50,3 +50,10 @@ COPY --from=fetch-nvidia \
 RUN apt-get install "/tmp/${NVIDIA_COMMON_DEB?}" \
   "/tmp/${NVIDIA_GL_DEB?}" \
   "/tmp/${NVIDIA_COMPUTE_DEB?}"
+
+# install cuda sdk
+RUN wget https://developer.download.nvidia.com/compute/cuda/11.6.0/local_installers/cuda-repo-debian11-11-6-local_11.6.0-510.39.01-1_amd64.deb
+RUN dpkg -i cuda-repo-debian11-11-6-local_11.6.0-510.39.01-1_amd64.deb
```
Comment (GMNGeoffrey) [inline build_tools/docker/nvidia/Dockerfile:56]: When writing scripts please prefer long options as they provide better documentation as to the intent (bundled into the suggestion below)

**Comment by ThomasRaoux** [comment]
@GMNGeoffrey, let me know if you have any more comments. Otherwise it should be good to go

File: build_tools/kokoro/gcp_ubuntu/cmake/linux/x86-turing/build.sh [hunk 14]
```diff
@@ -33,10 +33,44 @@ nvidia-smi || true
...
+cd ${ROOT_DIR?}
+rm -rf build/
+
+CMAKE_BIN=${CMAKE_BIN:-$(which cmake)}
+
+"$CMAKE_BIN" --version
+ninja --version
+
+cd ${ROOT_DIR?}
+
```
Comment (GMNGeoffrey) [inline build_tools/kokoro/gcp_ubuntu/cmake/linux/x86-turing/build.sh:50]: We just did this above

```suggestion
```

File: build_tools/kokoro/gcp_ubuntu/cmake/linux/x86-turing/build.sh [hunk 15]
```diff
@@ -33,10 +33,44 @@ nvidia-smi || true
...
+CMAKE_ARGS=(
+  "-G" "Ninja"
+  # Let's make linking fast
+  "-DIREE_ENABLE_LLD=ON"
+  "-DCMAKE_BUILD_TYPE=RelWithDebInfo"
+
+  "-DIREE_BUILD_PYTHON_BINDINGS=ON"
+
+  # Enable assertions.
```
Comment (GMNGeoffrey) [inline build_tools/kokoro/gcp_ubuntu/cmake/linux/x86-turing/build.sh:62]: This comment is redundant. The option is self-documenting

```suggestion
```

File: build_tools/docker/base/Dockerfile [hunk 16]
```diff
@@ -50,8 +50,8 @@ WORKDIR /install-cmake
 # to get the /usr/share path.
 # See https://github.com/moby/moby/issues/41383
 ARG CMAKE_MAJOR_VERSION=3
-ARG CMAKE_MINOR_VERSION=16
-ARG CMAKE_PATCH_VERSION=3
+ARG CMAKE_MINOR_VERSION=22
```
Comment (GMNGeoffrey) [inline build_tools/docker/base/Dockerfile:53]: Ergh, missed this, but this should've been upgraded to the *minimum* version on the maximum. I'll send a follow-up to drop it back to 3.17

Comment (ThomasRaoux) [reply build_tools/docker/base/Dockerfile:53]: ah this is in order to validate that we support the minimum version? I was thinking upgrading directly to 22 was simpler, but I didn't think about how to validate the minimum version. Sorry about that.

**Pull request merged**